File: primitives/vector3.ts
import { Primitive } from "./primitive.interface";

export class Vector3 implements Primitive {
	static readonly ZERO = new Vector3(0, 0, 0);
	static readonly RIGHT = new Vector3(1, 0, 0);
	static readonly UP = new Vector3(0, 1, 0);
	static readonly LEFT = new Vector3(-1, 0, 0);
	static readonly DOWN = new Vector3(0, -1, 0);
	static readonly FORWARD = new Vector3(0, 0, 1);
	static readonly BACKWARD = new Vector3(0, 0, -1);

	get length() {
		return Math.hypot(this.x, this.y, this.z);
	}

	constructor (
		public x: number,
		public y: number,
		public z: number
	) {}

	add(other: Vector3): Vector3 {
		return new Vector3(this.x + other.x, this.y + other.y, this.z + other.z);
	}

	subtract(other: Vector3): Vector3 {
		return new Vector3(this.x - other.x, this.y - other.y, this.z - other.z);
	}

	multiply(factor: number): Vector3 {
		return new Vector3(this.x * factor, this.y * factor, this.z * factor);
	}

	scalarProduct(other: Vector3): number {
		return this.x * other.x + this.y * other.y + this.z * other.z;
	}

	invert(): Vector3 {
		return new Vector3(-this.x, -this.y, -this.z);
	}

	normalize(): Vector3 {
		const length = this.length;

		if (length == 0) {
			return new Vector3(0, 0, 0);
		}

		return new Vector3(this.x / length, this.y / length, this.z / length);
	}

	/**
	 * Calculates the angle between two vectors
	 * @param other Other vector
	 * @returns Angle in radiants
	 */
	angle(other: Vector3): number {
		return Math.acos(this.scalarProduct(other) / (this.length * other.length));
	}

	clone(): Vector3 {
		return new Vector3(this.x, this.y, this.z);
	}

	toString(): string {
		return `Vector3[${this.x}, ${this.y}, ${this.z}]`;
	}
}

File: src/primitives/vector3.rs
use std::fmt::{Display, Formatter, Result};
use std::ops::{Add, AddAssign, Sub, SubAssign, Mul, MulAssign, Div, DivAssign, Neg};

#[derive(Debug, Clone, Copy, PartialEq)]
pub struct Vector3 {
	pub x: f64,
	pub y: f64,
	pub z: f64
}

#[allow(dead_code)]
impl Vector3 {
	pub const ZERO: Vector3 = Vector3 { x: 0.0, y: 0.0, z: 0.0 };
	pub const RIGHT: Vector3 = Vector3 { x: 1.0, y: 0.0, z: 0.0 };
	pub const UP: Vector3 = Vector3 { x: 0.0, y: 1.0, z: 0.0 };
	pub const LEFT: Vector3 = Vector3 { x: -1.0, y: 0.0, z: 0.0 };
	pub const DOWN: Vector3 = Vector3 { x: 0.0, y: -1.0, z: 0.0 };
	pub const FORWARD: Vector3 = Vector3 { x: 0.0, y: 0.0, z: 1.0 };
	pub const BACKWARD: Vector3 = Vector3 { x: 0.0, y: 0.0, z: -1.0 };
}

impl Vector3 {
	pub fn new(x: f64, y: f64, z: f64) -> Vector3 {
		Self { x, y, z }
	}

	pub fn length(&self) -> f64 {
		(self.x * self.x + self.y * self.y + self.z * self.z).sqrt()
	}

	pub fn normalize(&self) -> Vector3 {
		let length = self.length();

		if length == 0.0 {
			Vector3::ZERO
		} else {
			*self / length
		}
	}

	pub fn cross(&self, other: Vector3) -> Vector3 {
		Vector3::new(
			self.y * other.z - self.z * other.y,
			self.z * other.x - self.x * other.z,
			self.x * other.y - self.y * other.x
		)
	}

	pub fn dot(&self, other: &Vector3) -> f64 {
		self.x * other.x + self.y * other.y + self.z * other.z
	}

	pub fn angle(&self, other: &Vector3) -> f64 {
		(self.dot(other) / (self.length() * other.length())).acos()
	}
}

impl Add for Vector3 {
	type Output = Vector3;

	fn add(self, other: Vector3) -> Vector3 {
		Vector3::new(self.x + other.x, self.y + other.y, self.z + other.z)
	}
}

impl AddAssign for Vector3 {
	fn add_assign(&mut self, other: Vector3) {
		*self = *self + other;
	}
}

impl Sub for Vector3 {
	type Output = Vector3;

	fn sub(self, other: Vector3) -> Vector3 {
		Vector3::new(self.x - other.x, self.y - other.y, self.z - other.z)
	}
}

impl SubAssign for Vector3 {
	fn sub_assign(&mut self, other: Vector3) {
		*self = *self - other;
	}
}

impl Mul for Vector3 {
	type Output = Vector3;

	fn mul(self, other: Vector3) -> Vector3 {
		Vector3::new(self.x * other.x, self.y * other.y, self.z * other.z)
	}
}

impl MulAssign for Vector3 {
	fn mul_assign(&mut self, other: Vector3) {
		*self = *self * other;
	}
}

impl Mul<f64> for Vector3 {
	type Output = Vector3;

	fn mul(self, factor: f64) -> Vector3 {
		Vector3::new(self.x * factor, self.y * factor, self.z * factor)
	}
}

impl MulAssign<f64> for Vector3 {
	fn mul_assign(&mut self, other: f64) {
		*self = *self * other;
	}
}

impl Div for Vector3 {
	type Output = Vector3;

	fn div(self, other: Vector3) -> Vector3 {
		Vector3::new(self.x / other.x, self.y / other.y, self.z / other.z)
	}
}

impl DivAssign for Vector3 {
	fn div_assign(&mut self, other: Vector3) {
		*self = *self / other;
	}
}

impl Div<f64> for Vector3 {
	type Output = Vector3;

	fn div(self, factor: f64) -> Vector3 {
		Vector3::new(self.x / factor, self.y / factor, self.z / factor)
	}
}

impl DivAssign<f64> for Vector3 {
	fn div_assign(&mut self, other: f64) {
		*self = *self / other;
	}
}

impl Neg for Vector3 {
	type Output = Vector3;

	fn neg(self) -> Vector3 {
		Vector3::new(-self.x, -self.y, -self.z)
	}
}

impl Display for Vector3 {
	fn fmt(&self, f: &mut Formatter<'_>) -> Result {
		write!(f, "Vector3[{}, {}, {}]", self.x, self.y, self.z)
	}
}
